README.md
# Demo of the game of "Set"

[![Source](https://img.shields.io/badge/source-skyzyx/set–game–demo-blue.svg?style=flat-square)](https://github.com/skyzyx/set-game-demo)
[![Downloads](https://img.shields.io/pypi/dm/skyzyx-set-game-demo.svg?style=flat-square)](https://github.com/skyzyx/set-game-demo/releases)
[![Release](https://img.shields.io/github/release/skyzyx/set-game-demo.svg?style=flat-square)](https://github.com/skyzyx/set-game-demo/releases)
[![Pypi Release](https://img.shields.io/pypi/v/skyzyx-set-game-demo.svg?style=flat-square)](https://pypi.python.org/pypi/skyzyx-set-game-demo)
[![Open Issues](http://img.shields.io/github/issues/skyzyx/set-game-demo.svg?style=flat-square)](https://github.com/skyzyx/set-game-demo/issues)
[![Build Status](http://img.shields.io/travis/skyzyx/set-game-demo/master.svg?style=flat-square)](https://travis-ci.org/skyzyx/set-game-demo)
[![Implementation](https://img.shields.io/pypi/implementation/skyzyx-set-game-demo.svg?style=flat-square)](https://python.org)
[![Python Versions](https://img.shields.io/pypi/pyversions/skyzyx-set-game-demo.svg?style=flat-square)](https://python.org)
[![Package Format](https://img.shields.io/pypi/format/skyzyx-set-game-demo.svg?style=flat-square)](http://pythonwheels.com)
[![Stability](https://img.shields.io/pypi/status/skyzyx-set-game-demo.svg?style=flat-square)](https://pypi.python.org/pypi/skyzyx-set-game-demo)
[![Coverage Status](http://img.shields.io/coveralls/skyzyx/set-game-demo/master.svg?style=flat-square)](https://coveralls.io/r/skyzyx/set-game-demo?branch=master)
[![Code Climate](http://img.shields.io/codeclimate/github/skyzyx/set-game-demo.svg?style=flat-square)](https://codeclimate.com/github/skyzyx/set-game-demo)
[![Code Quality](http://img.shields.io/scrutinizer/g/skyzyx/set-game-demo.svg?style=flat-square)](https://scrutinizer-ci.com/g/skyzyx/set-game-demo)
[![License](https://img.shields.io/github/license/skyzyx/set-game-demo.svg?style=flat-square)](https://github.com/skyzyx/set-game-demo/blob/master/LICENSE.md)
[![Author](http://img.shields.io/badge/author-@skyzyx-blue.svg?style=flat-square)](https://github.com/skyzyx)

"Set" is a card game where a group of players try to identify a _Set_ of cards from those placed face-up on a table.

This project uses [Semantic Versioning](http://semver.org) for managing backwards-compatibility.

* [API Reference](https://skyzyx.github.io/set-game-demo/)


## The Game

### Core Concepts

* Each _Card_ has 4 _Properties_: color, shape, shading, and number.
* The _Deck_ is a collection of all of the _Cards_.
* The _Board_ is a subset of the _Deck_, containing only the cards that are currently in-play.
* A _Set_ is a collection of 3 cards which meet certain criteria (discussed below). When a _Set_ is found in-play on the _Board_, the _Set_ is removed from play and logged as such.
* The _Game_ encapsulates all of these concepts and keeps track of them.

### Game Rules

Each _Card_ has an image on it with 4 orthogonal attributes:

* Color (red, green, or purple)
* Shape (diamond, squiggle, or oval)
* Shading (solid, empty, or striped)
* Number (one, two, or three)

Three _Cards_ are a part of a _Set_ if, for each _Property_, the values are all the same or all different.

For example:

* The _Cards_ "two red solid squiggles", "one green solid diamond", "three purple solid ovals" would make up a
  _Set_. (number, shape, and color are different, shading is the same)
* The _Cards_ "two red solid squiggles", "one green solid squiggles", "three purple solid ovals" would not make up a
  _Set_, because shape is the same on two _Cards_, but different on the third.
* A _Game_ of "Set" starts by dealing 12 _Cards_, face-up. When a player sees three _Cards_ that make up a _Set_,
  they yell "Set!" and grab the _Cards_. New _Cards_ are dealt from the _Deck_ to replace them.
* If no player can find a _Set_, three more _Cards_ are dealt (to make 15, then 18, then 21...)
* The _Game_ is over when there are no _Cards_ left in the _Deck_, and no _Sets_ left on the table. The player with
  the most _Sets_ wins.

### Game Requirements

Your task is to model the _Game_ in code, and implement the following methods:

* A method that takes three _Cards_, and determines whether the three _Cards_ make a _Set_.
* A method that given a _Board_ of _Cards_, will either find a _Set_, or determine that there are no _Sets_ on the
  table.
* A method that will play an entire _Game_ of "Set", from beginning to end, and return a list of each valid _Sets_
  you removed from the _Board_.

For this last method, there will be multiple correct solutions, but any valid list of _Sets_ is fine.

### Assumptions

> _“Three cards are a part of a set if, for each property, the values are all the same or all different.”_

This is phrased ambiguously, and the examples given lead me to believe that the following is a more specific description of the rules.

* Take 3 cards and look at each of their properties one-by-one.
* If all cards have a different value for that property OR all cards have the same value for that property, then it _may_ be a set.
* If _any_ properties of step 2 fail the test, then the group is not a set.

### Problem Parameters

* This problem uses mathematical _combinations_ (as opposed to _permutations_). This results in 81 combinations (`3^4`).
* Any failure of being a _Set_ means that the group is not a set, so fail as early as possible and move-on.


### Logic

(Whereas “_Combination_” refers to the [mathematical concept](https://en.wikipedia.org/wiki/Combination).)

1. Create the deck of available cards by ensuring that every card is unique, and that all _Combinations_ of properties are represented. Also, shuffle the deck by default.
1. Deal 12 cards to the _Board_.
1. Calculate all possible _Combinations_ of the cards on the _Board_, in groups of 3.
1. Iterate over each _Combination_, applying logic to determine whether or not this _Combination_ represents a _Set_.
1. Collect the _Sets_ by removing the _Cards_ which are determined to be part of a _Set_.
1. When no more _Sets_ can be found, deal another 3 _Cards_ from the _Deck_.
1. Repeat steps 3–6 until the _Deck_ is empty.

## Requirements

* Python 2.7, 3.3+, Pypy
* Pip
* VirtualEnv is recommended, but not required


## Installation

```bash
# Install from Pypi
pip install skyzyx-set-game-demo

# Install from local code
pip install -e .
```

And either include it in your scripts:

```python
from set_game_demo import SetGame
```

…or run it from the command line.

```bash
# Application help
set-game-demo -h
```

## Usage/Examples

From the Python REPL or a Python script…

```python
from __future__ import print_function
from set_game_demo import SetGame

# Initialize the game.
game = SetGame()

# Chatty, interactive version of the game.
game.play()

# Quiet version of the game. Good for code.
discovered, sets = game.play_quiet()
print("Sets discovered: {}".format(discovered))
for set in sets:
    game.display_cards(set)
```

From the Terminal…

```bash
# Chatty, interactive version of the game.
set-game-demo

# Quiet version of the game.
set-game-demo --quiet
```

## Known Issues

* In a final release, it would be wise to update the `requirements.txt` to allow for ranges of known-good versions instead of locking to one specific version.
    * Conversely, if this is the sole project running in this virtual environment, locking to a specific known-good version ensures fewer version-compatibility issues.


## Future Improvements

* Update the `test_deal` unit test to verify that we do not attempt to deal a larger number of cards than the deck contains (couldn't quite figure out the right way to call `assertRaises()` from the `unittest` package through the `nose2` interface).
* Support multiple _Players_ who can collect sets and compete for scores.


## Development

* You can develop in any supported version of Python.

* Using [pyenv] to manage your Pythons is _highly-recommended_. Testing locally **depends** on it.

* Install [VirtualEnv] for your development environment, and _activate_ the environment.

  ```bash
  pip install virtualenv
  virtualenv .vendor
  source .vendor/bin/activate
  ```

* Install the `requirements.txt`.

  ```bash
  pip install -r requirements.txt
  ```

* When you make changes, make sure that you run the linter and fix anything that's broken.

  ```bash
  make lint
  ```


## Testing

We use [tox] to handle local testing across multiple versions of Python. We install multiple versions of Python at a time with [pyenv].

Testing occurs against the following versions:

* Python 2.7
* Python 3.3
* Python 3.4
* Python 3.5
* Python 3.6
* Python 3.7
* Pypy
* Pypy3

To begin…

1. Install [pyenv] on your own before running tests.

1. You need to install all of the supported versions of Python. (This will take a while.) If you would prefer to install your own copies of the supported Python versions (listed above), feel free to manage them yourself.

   ```bash
   pyenv install 3.7.1 && \
   pyenv install 3.6.7 && \
   pyenv install 3.5.6 && \
   pyenv install 3.4.9 && \
   pyenv install 3.3.7 && \
   pyenv install 2.7.15 && \
   pyenv install pypy-5.7.1 && \
   pyenv install pypy3.5-6.0.0 && \
   pyenv rehash && \
   eval "$(pyenv init -)" && \
   pyenv global system 2.7.15 3.3.7 3.4.9 3.5.6 3.6.7 3.7.1 pypy-5.7.1 pypy3.5-6.0.0
   ```

   To verify that the installation and configuration were successful, you can run `pyenv versions`. You should see a `*` character in front of every version that we just installed.

   ```bash
   $ pyenv versions
   * system (set by ~/.pyenv/version)
   * 2.7.12 (set by ~/.pyenv/version)
   * 3.3.6 (set by ~/.pyenv/version)
   * 3.4.5 (set by ~/.pyenv/version)
   * 3.5.2 (set by ~/.pyenv/version)
   * 3.6.0b1 (set by ~/.pyenv/version)
   * pypy-5.3.1 (set by ~/.pyenv/version)
   * pypy3-2.4.0 (set by ~/.pyenv/version)
   ```

1. The following command will package-up your module and install it locally, then run `nose2` to execute the tests in the _default system Python_.

   ```bash
   make test
   ```


1. After you've run that, you can then execute the tests in all supported versions of Python with the following:

   ```bash
   tox
   ```

## API Reference

### Building local docs

```bash
make docs
open docs/set_game_demo/index.html
```

### Building and pushing docs

```bash
make pushdocs
```

Docs can be viewed at <https://skyzyx.github.io/set-game-demo/>.


## Deploying

1. The `Makefile` (yes, `Makefile`) has a series of commands to simplify the development and deployment process.
1. Also install [Chag]. This is used for managing the `CHANGELOG` and annotating the Git release correctly.

### Updating the CHANGELOG

Make sure that the `CHANGELOG.md` is human-friendly. See http://keepachangelog.com if you don’t know how.

### `make`

Running `make` by itself will show you a list of available sub-commands.

```bash
$ make
all
buildpip
clean
docs
lint
pushdocs
pushpip
readme
tag
test
version
```

### `make readme`

If you make changes to `README.md`, then this will use [Pandoc] to output a `README.rst` file in the [reStructuredText] format used by [distutils], [Sphinx] and most of the Python community.

You must have [Pandoc] installed on your local system.

> **NOTE:** Initial install via `brew install pandoc` takes about 8–10 hours. Updates are much faster. [Using the installer](https://github.com/jgm/pandoc/releases) is **much** faster for initial installation, but updates are entirely manual.

### `make version`

Sets the version number that will be used by other `make` tasks related to packaging and bundling.

### `make tag`

This will make sure that the `CHANGELOG.md` is properly datestamped, add the CHANGELOG contents to the Git commit message, commit them, then create a Git commit which can be pushed upstream.

### `make buildpip`

This will bundle-up your package in preparation for uploading to [Pypi].

### `make pushpip`

This will take your bundled package and upload it securely to [Pypi] using the `twine` package.

### Drafting a GitHub release

1. Go to https://github.com/skyzyx/set-game-demo/tags
1. Find the new tag that you just pushed. Click the ellipsis (`…`) to see the commit notes. Copy these.
1. To the right, choose _Add release notes_. Your _Tag version_ should be pre-filled.
1. The _Release title_ should match your _Tag version_.
1. Inside _Describe this release_, paste the notes that you copied on the previous page.
1. Choose _Publish release_.
1. Your release should now be the latest. https://github.com/skyzyx/set-game-demo/releases/latest


## Contributing
Here's the process for contributing:

1. Fork this project to your GitHub account.
2. Clone your GitHub copy of the repository into your local workspace.
3. Write code, fix bugs, and add tests with 100% code coverage.
4. Commit your changes to your local workspace and push them up to your GitHub copy.
5. You submit a GitHub pull request with a description of what the change is.
6. The contribution is reviewed. Maybe there will be some banter back-and-forth in the comments.
7. If all goes well, your pull request will be accepted and your changes are merged in.


## Authors, Copyright & Licensing

* Copyright (c) 2016 [Ryan Parman](http://github.com/skyzyx)

See also the list of [contributors](https://github.com/skyzyx/set-game-demo/graphs/contributors) who participated in this project.

Licensed for use under the terms of the [Apache 2.0] license.

  [Apache 2.0]: http://opensource.org/licenses/Apache-2.0
  [Chag]: https://github.com/mtdowling/chag
  [distutils]: https://docs.python.org/3/library/distutils.html
  [Pandoc]: http://pandoc.org
  [pyenv-virtualenvwrapper]: https://github.com/yyuu/pyenv-virtualenvwrapper
  [pyenv]: https://github.com/yyuu/pyenv
  [reStructuredText]: http://docutils.sourceforge.net/rst.html
  [Sphinx]: http://www.sphinx-doc.org
  [tox]: https://tox.readthedocs.io
  [VirtualEnv]: https://virtualenv.pypa.io/en/stable/
  [Pypi]: https://pypi.python.org/pypi


CHANGELOG.md
# CHANGELOG

## 1.0.1 - 2018-11-21

* Updated the dependency versions.
* Updated the docs for running tests.

## 1.0.0 - 2016-09-26

*  Initial release.


set_game_demo/__init__.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

"""
A simple demo of the game of "Set".

Copyright (c) 2016 [Ryan Parman](https://github.com/skyzyx).

<http://opensource.org/licenses/Apache2.0>
"""

from __future__ import print_function
import argparse
import collections
import itertools
import random
import six
from prettytable import PrettyTable


class SetGame(object):
    """
    "Set" is a card game where a group of players try to identify a _Set_ of cards from those placed face-up on a table.

    **Game Rules:**

    Each _Card_ has an image on it with 4 orthogonal attributes:

    * Color (red, green, or purple)
    * Shape (diamond, squiggle, or oval)
    * Shading (solid, empty, or striped)
    * Number (one, two, or three)

    Three _Cards_ are a part of a _Set_ if, for each _Property_, the values are all the same or all different.

    For example:

    * The _Cards_ "two red solid squiggles", "one green solid diamond", "three purple solid ovals" would make up a
      _Set_. (number, shape, and color are different, shading is the same)
    * The _Cards_ "two red solid squiggles", "one green solid squiggles", "three purple solid ovals" would not make up a
      _Set_, because shape is the same on two _Cards_, but different on the third.
    * A _Game_ of "Set" starts by dealing 12 _Cards_, face-up. When a player sees three _Cards_ that make up a _Set_,
      they yell "Set!" and grab the _Cards_. New _Cards_ are dealt from the _Deck_ to replace them.
    * If no player can find a _Set_, three more _Cards_ are dealt (to make 15, then 18, then 21...)
    * The _Game_ is over when there are no _Cards_ left in the _Deck_, and no _Sets_ left on the table. The player with
      the most _Sets_ wins.

    **Game Requirements:**

    Your task is to model the _Game_ in code, and implement the following methods:

    * A method that takes three _Cards_, and determines whether the three _Cards_ make a _Set_.
    * A method that given a _Board_ of _Cards_, will either find a _Set_, or determine that there are no _Sets_ on the
      table.
    * A method that will play an entire _Game_ of "Set", from beginning to end, and return a list of each valid _Sets_
      you removed from the _Board_.

    For this last method, there will be multiple correct solutions, but any valid list of _Sets_ is fine.

    **Compatibility:**

    Tested against the following Python versions:

    * Python 2.7
    * Python 3.3
    * Python 3.4
    * Python 3.5
    * Python 3.6 (beta)
    * Pypy (~2.7.10)
    * Pypy3 (~3.2.5)

    **Import:**

        from set_game_demo import SetGame

    """

    def __init__(self):
        """
        Constructs a new instance of this class.
        """

        self.colors = ["red", "green", "purple"]
        self.shapes = ["diamond", "squiggle", "oval"]
        self.shadings = ["solid", "empty", "striped"]
        self.numbers = ["one", "two", "three"]

        deck = []

        for color in self.colors:
            for shape in self.shapes:
                for shading in self.shadings:
                    for number in self.numbers:
                        card = {}
                        card["color"] = color
                        card["shape"] = shape
                        card["shading"] = shading
                        card["number"] = number

                        deck.append(card)

        random.shuffle(deck)

        self.deck = collections.deque(deck)
        self.board = []
        self.sets = []

    def deal(self, cards=12):
        """
        Deals a given number of cards from the top (front) of the deck.

        `cards (integer)`: The number of cards to deal out to the board. The default value is `12`.

        `return (cards[])`: A list (i.e., array) of objects representing cards.
        """

        out = []

        for _ in six.moves.range(cards):
            out.append(self.deck.popleft())

        return out

    def play(self):  # pragma: no cover
        """
        Play a (chatty) game of Set.

        `return (void)`
        """
        # pylint: disable=R0915

        # Introduction
        print("Welcome to a game of Set.")
        print()
        six.moves.input("=> Press any key to continue...")

        # First deal
        print()
        print("Dealing 12 cards onto the board.")
        self.board = self.deal(12)
        SetGame.display_cards(self.board)
        print()
        print("Cards on the board: {}".format(len(self.board)))
        print("Cards in the deck:  {}".format(len(self.deck)))
        print("Sets discovered:    {}".format(len(self.sets)))
        six.moves.input("=> Press any key to continue...")

        while len(self.deck) > 0:
            # Find sets
            print()
            sets = self.find_sets(self.board)
            print("Discovered {quantity}.".format(
                quantity=SetGame.plural(len(sets), "set", "sets")
            ))

            for i, sset in enumerate(sets):
                print()
                print("Set #{index}".format(index=(i + 1)))
                SetGame.display_cards(sset)

            print()
            print("These are the remaining {quantity} on the board.".format(
                quantity=SetGame.plural(len(self.board), "card", "cards")
            ))
            SetGame.display_cards(self.board)
            print()
            self.sets += sets
            six.moves.input("=> Press any key to continue...")

            # Deal 3 more cards
            print()
            self.board += self.deal(3)
            print("Dealing 3 more cards. You now have {quantity} on the board.".format(
                quantity=SetGame.plural(len(self.board), "card", "cards")
            ))
            SetGame.display_cards(self.board)
            print()
            print("Cards on the board: {}".format(len(self.board)))
            print("Cards in the deck:  {}".format(len(self.deck)))
            print("Sets discovered:    {}".format(len(self.sets)))
            six.moves.input("=> Press any key to continue...")

        # Find the very last set(s)
        print()
        sets = self.find_sets(self.board)
        print("Discovered {quantity}.".format(
            quantity=SetGame.plural(len(sets), "set", "sets")
        ))

        for i, sset in enumerate(sets):
            print()
            print("Set #{index}".format(index=(i + 1)))
            SetGame.display_cards(sset)

        print()
        print("These are the remaining {quantity} on the board.".format(
            quantity=SetGame.plural(len(self.board), "card", "cards")
        ))
        SetGame.display_cards(self.board)
        print()
        self.sets += sets
        six.moves.input("=> Press any key to continue...")

        # No more cards in the deck.
        print()
        print("There are no more cards left in the deck.")
        print()
        print("These are the {} that are left on the table.".format(
            SetGame.plural(len(self.board), "card", "cards")
        ))
        SetGame.display_cards(self.board)
        print()
        six.moves.input("=> Congratulations! You have completed the game. Press any key to see the results.")

        # Final score.
        print()
        print("You discovered {quantity}.".format(
            quantity=SetGame.plural(len(self.sets), "set", "sets")
        ))

        for i, sset in enumerate(self.sets):
            print()
            print("Set #{index}".format(index=(i + 1)))
            SetGame.display_cards(sset)

        # All done.
        print()
        print("Please play again!")
        print()

    def play_quiet(self):
        """
        Play a (quiet) game of Set.

        `return (tuple(integer, sets[]))`: Returns a tuple where the first item is the number of Sets discovered. The
            second item is the complete list of Sets.
        """
        self.board = self.deal(12)

        # Find sets and re-deal until we are out of cards.
        while len(self.deck) > 0:
            self.sets += self.find_sets(self.board)
            self.board += self.deal(3)

        # Find the very last set(s).
        self.sets += self.find_sets(self.board)

        # Return a tuple
        return (len(self.sets), self.sets)

    @staticmethod
    def display_cards(cards):  # pragma: no cover
        """
        Accepts an array of cards, and renders them for humans in a table to `stdout`.

        `cards (cards[])`: A list (i.e., array) of objects representing cards.

        `return (void)`
        """

        table = PrettyTable(["Number", "Color", "Shape", "Shading"])

        for card in cards:
            table.add_row([card["number"], card["color"], card["shape"], card["shading"]])

        print(table)

    @staticmethod
    def is_a_set(card1, card2, card3):
        """
        Determines whether a group of 3 cards is a _set_ or not.

        `card1 (card)`: A card object to compare against other card objects.

        `card2 (card)`: A card object to compare against other card objects.

        `card3 (card)`: A card object to compare against other card objects.

        `return (boolean)`: Whether or not the given cards represent a _Set_. A value of `True` means that the group is
            a set. A value of `False` means that the group is _not_ a set.
        """

        for attribute in ["color", "shape", "shading", "number"]:
            if (SetGame.__all_unique([
                    card1[attribute],
                    card2[attribute],
                    card3[attribute],
                ]) is False

                    and SetGame.__all_same([
                        card1[attribute],
                        card2[attribute],
                        card3[attribute],
                    ]) is False):

                return False

        # Did we actually make it all the way here?
        return True

    @staticmethod
    def find_sets(board):
        """
        Given a _Board_ of cards, determines whether or not it contains a _Set_.

        `board (cards[])`: A list (i.e., array) of Cards that are on the Board.

        `return (sets[])`: A list (i.e., array) of Sets. Each Set contains 3 Cards.
        """

        sets = []

        # Calculate the initial set of combinations.
        combinations = itertools.combinations(six.moves.range(len(board)), 3)

        # Run until we explicitly break.
        while True:

            # Grab the combination.
            combination = next(combinations, None)

            # As long as we didn't get a `None` back...
            while combination is not None:

                # Check to see if the three cards we got back for this combination are a Set.
                if SetGame.is_a_set(board[combination[0]], board[combination[1]], board[combination[2]]) is True:

                    # If so, save them.
                    sets.append([board[combination[0]], board[combination[1]], board[combination[2]]])

                    # After we save the Set, remove the cards from the Board.
                    # Python list indexes collapse automatically, so remove from the end first.
                    del board[combination[2]]
                    del board[combination[1]]
                    del board[combination[0]]

                    # This means that we now need to recalculate the Board, and start our loop over again.
                    combinations = itertools.combinations(six.moves.range(len(board)), 3)
                    break

                else:
                    # Move on to the next combination in the list.
                    combination = next(combinations, None)
                    continue

            # If we have reached the end of the list of combinations, quit.
            if combination is None:
                break

        # Return all matching sets.
        return sets

    @staticmethod
    def __all_unique(arr):
        """
        Determines whether or not a `List` of values are all unique.

        `arr (scalar[])`: A list (i.e., array) of scalar values to compare.

        `return (boolean)`: Whether or not a `List` of values are all different. A value of `True` means that the values
            are all unique. A value of `False` means that the values are not all unique.
        """

        return len(set(arr)) == len(arr)

    @staticmethod
    def __all_same(arr):
        """
        Determines whether or not a `List` of values are all identical.

        `arr (scalar[])`: A list (i.e., array) of scalar values to compare.

        `return (boolean)`: Whether or not a `List` of values are all identical. A value of `True` means that the values
            are all identical. A value of `False` means that the values are not all identical.
        """

        return len(set(arr)) == 1

    @staticmethod
    def plural(count, singular, plural):
        """
        Determines whether a response should be singular or plural, depending on the number of items being referenced.

        `count (integer)`: The number of items to adjust language for.

        `singular (string)`: The singular version of the phrasing.

        `plural (string)`: The plural version of the phrasing.

        `return (string)`: A properly-formatted string.
        """

        return "{count} {word}".format(
            count=count,
            word=singular
        ) if count == 1 else "{count} {word}".format(
            count=count,
            word=plural
        )

# ------------------------------------------------------------------------------


def main():  # pragma: no cover
    """
    This function is run when the script is executed from the command-line.
    """

    # Available CLI flags.
    parser = argparse.ArgumentParser(
        description="Play a game of Set.",
    )

    parser.add_argument(
        "-q", "--quiet",
        dest="quiet",
        action="store_true",
        help="By default, the game will play in chatty, interactive mode. "
        "This will enable a quieter, results-only mode.")

    parser.set_defaults(quiet=False)
    flags = parser.parse_args()
    game = SetGame()

    if flags.quiet:
        discovered, sets = game.play_quiet()
        print("You discovered {quantity}.".format(
            quantity=SetGame.plural(discovered, "set", "sets")
        ))

        for i, sset in enumerate(sets):
            print()
            print("Set #{index}".format(index=(i + 1)))
            SetGame.display_cards(sset)

    else:
        game.play()


if __name__ == "__main__":  # pragma: no cover
    main()
